### DISCUSSION TITLE
add_dll_directory to specify dll paths on windows

### DISCUSSION THREAD

**Discussion by james-thunes** [pytest-dev/pytest#10692] [Q&A] [answered]
hello,

I've been working on update of a (somewhat complex) program from python 3.7 to 3.10. Our application uses a swig to generate a number of pyd files that reference dlls on windows. Due to quirks of our build, the pyd and dll files are not in the same directory. In python 3.8+, this causes some issues as our previous method of adding the paths to `PATH` on windows is no longer valid. Python no longer uses PATH to search for dlls. Instead, you must explicitly add the paths using `os.add_dll_directory()`.

This is fine for our normal executable. I can add the directories as required in the main python module and python is able to import the dlls in the pyd libraries.

However, for pytest, I'm seeing errors such as these when I attempt to use existing unittests with python 3.10:
```
 ______________________________ ERROR collecting kernel/units/unittest/TestUnitsChecks.py ______________________________
ImportError while importing test module 'C:\ANSYSDev\develop\SystemCoupling\PyLib\kernel\units\unittest\TestUnitsChecks.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
..\..\Core_Dependencies\CPython\3_10\winx64\Release\python\lib\importlib\__init__.py:126: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
kernel\units\__init__.py:9: in <module>
    import pyExt.Units
..\runTime\winx64\bin\pyExt\Units.py:13: in <module>
    from . import _Units
E   ImportError: DLL load failed while importing _Units: The specified module could not be found.
```

The pyd file can't find the dll dependency. Adding `os.add_dll_directory()` to the unittest module doesn't allow for the dll to be found. The dll is found if I add `os.add_dll_directory()` to `kernel\units\__init__.py`, but that's not ideal as I do not want that call to be in the module when not running unittests.

Does anybody know where I should define the paths such that pytest can load the module?

Reading the doc, I do see that there's a `pythonpath` config option, but that only adds paths to `PYTHONPATH` for python module imports, not dll imports.

Just to confirm, all my unittests do pass if I copy my dlls into the same directory as the pyd files. I can update my builds to do this, but I was hoping to keep my updates to a minimum.

**Comment by nicoddemus** [comment, marked as answer]
Hi,

Try adding your `os.add_dll_directory` call to your top-most `conftest.py` file. pytest loads the top-most `conftest.py` file before starting to load the test modules, so this should work.

**Reply by james-thunes** [reply]
that seems to have done the trick. Thanks!
